src/relay/vm/compiler.cc

```diff
@@ -0,0 +1,611 @@
+        RELAY_LOG(INFO)
+          << "VisitExpr_: generating invoke for "
+          << global->name_hint
+          << " with func_index="
+          << it->second;
+
+        auto func = this->context->module->Lookup(global);
+        if (IsClosure(func)) {
+          auto arity = func->params.size();
+          std::vector<size_t> free_var_registers;
+          for (size_t i = 0; i < arity; ++i) {
+            free_var_registers.push_back(var_register_map.at(func->params[i]));
+          }
+          Emit(AllocClosure(it->second, arity, free_var_registers, NewRegister()));
+        } else {
+          Emit(Invoke(it->second, args_registers, NewRegister()));
+        }
+      } else if (auto constructor_node = op.as<ConstructorNode>()) {
+        auto constructor = GetRef<Constructor>(constructor_node);
+        auto tag = GetConstructorTag(constructor);
+        Emit(AllocDatatype(tag, call_node->args.size(), args_registers, NewRegister()));
+      } else if (auto var_node = op.as<VarNode>()) {
+        VisitExpr(GetRef<Var>(var_node));
+        Emit(InvokeClosure(last_register, args_registers, NewRegister()));
+      } else {
+        LOG(FATAL) << "unsupported case in vm compiler: " << op;
+      }
+    }
+
+    size_t GetConstructorTag(tvm::relay::Constructor constructor) {
+      auto it = this->context->tag_map.find(constructor);
+      if (it != this->context->tag_map.end()) {
+        return it->second;
+      } else {
+        auto tag = this->context->tag_map.size();
+        this->context->tag_map[constructor] = tag;
+        this->context->tag_index_map[tag] = constructor;
+        return tag;
+      }
+    }
+
+    void VisitExpr_(const FunctionNode* func_node) {
+      if (!func_node->IsPrimitive()) {
+        LOG(FATAL) << "local functions should have been removed by lambda lifting"
+                  << AsText(GetRef<Function>(func_node), false);
+      }
+    }
+
+    void CompileClosure(const Function& func) {
+      // We first layout the function arguments.
+      auto inner_func = Downcast<Function>(func->body);
+
+      size_t i = 0;
+      for (auto param : inner_func->params) {
+        auto arg_register = NewRegister();
+        CHECK_EQ(i, arg_register);
+        var_register_map.insert({ param, arg_register });
+        i++;
+      }
+
+      // We then assign register num to the free variables
+      for (auto param : func->params) {
+        auto arg_register = NewRegister();
+        CHECK_EQ(i, arg_register);
+        var_register_map.insert({ param, arg_register });
+        i++;
+      }
+
+      // We will now process the body like normal.
+      this->VisitExpr(inner_func->body);
+    }
+
+    void Compile(const Function& func) {
+      AsText(func, false);
+
+      // We need to generate code specially for lifted closures.
+      if (IsClosure(func)) {
+        CompileClosure(func);
+        return;
+      }
+
+      for (size_t i = 0; i < func->params.size(); ++i) {
```

Comment:
Move this code above `IsClosure` and delete the same code in `CompileClosure` since both `Function` and `Closure` have free vars.